Fix error applying `ignore_malformed` to boolean values

cbuescher opened the pull request:
The `ignore_malformed` option currently works on numeric fields only when the
bad value isn't a string value but not if it is a boolean. In this case we get a
parsing error from the xContent parser which we need to catch in addition to the
field mapper.

Closes #11498

Commit:
Fix error applying `ignore_malformed` to boolean values

The `ignore_malformed` option currently works on numeric fields only when the
bad value isn't a string value but not if it is a boolean. In this case we get a
parsing error from the xContent parser which we need to catch in addition to the
field mapper.

Closes #11498

elasticmachine commented:
Pinging @elastic/es-search

jtibshirani commented on server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:
```diff
@@ -218,45 +220,39 @@ protected void doTestDecimalCoerce(String type) throws IOException {
-                        .endObject()),
-                XContentType.JSON));
-        MapperParsingException e = expectThrows(MapperParsingException.class, runnable);
+            Object malformedValue = randomBoolean() ? "a" : false;
```
Instead of using randomization, it might be good to always test both values through a for loop. We just added some guidance to our testing docs around avoiding randomized testing for coverage: https://github.com/elastic/elasticsearch/blob/master/TESTING.asciidoc#bad-practices

cbuescher replied:
Sure

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -1042,7 +1044,7 @@ protected void parseCreateField(ParseContext context, List<IndexableField> field
             try {
                 numericValue = fieldType().type.parse(parser, coerce.value());
-            } catch (IllegalArgumentException e) {
+            } catch (IllegalArgumentException | JsonParseException e) {
```
I haven't tested this, but I wanted to check that this will not introduce too much leniency? In particular, I'm wondering what happens if we try to pass a JSON object to a numeric field with `ignore_malformed` enabled.

cbuescher replied:
The way the parser works, anything that's not a number or string token after the field name should throw an error. If "swallowing" that token with ignore_malformed leads to unparseable xcontent afterwards we will still throw an error. I think that's what should happen.

jtibshirani replied:
Thanks for clarifying. What would you think about adding a check whether the token is a value, as in if (ignoreMalformed.value() && parser.currentToken().isValue()) { ... }? That way a user would get a clear, localized error like "Current token (START_OBJECT) not numeric", as opposed to a more difficult one to track down like "Malformed content, found extra data after parsing: END_OBJECT".

cbuescher replied:
The problem with that is that the code actually throwing the exception is in the jackson library. The only way I could think of to do this is either have some logic taking apart the exception or pre-checking the token in NumberFieldMapper. I'd try to avoid the former, was the later what you had in mind? We'd need to test for (and accept) all value tokens here basically?

jtibshirani reviewed (commented):
This looks good to me overall, I left a question to make sure I understand the approach.

Commit:
iter

Commit:
Adressing comments

cbuescher commented:
@jtibshirani thanks for the review, I added the check for the token type you suggested and added a test where the malformed field value is an object. The current implementation will throw erros in numeric fields `foo` like:
```
{
   foo: { "someField" : "someValue" } 
}
```
regardless of the `ignore_malformed` setting. Not sure if that is what the user would expect here. As an alternative we could skip the whole object in case `ignore_malformed` is true. What do you think is the more intuitive behaviour in these edge cases?

jtibshirani reviewed (approved):
Thanks @cbuescher, this looks good to me. My intuition is that the whole object should be skipped, but it might make sense to address the object case in a future PR since there is a separate issue for it with quite a bit of discussion (#12366).

cbuescher commented:
@jtibshirani thanks for the review, and +1 to tackling the object case as a follow up, e.g. when tackling the issue you linked.
